server/schedule/operator/builder.go

```diff
@@ -0,0 +1,596 @@
+// will allocate a new peer ID later.
+func (b *Builder) AddPeer(p *metapb.Peer) *Builder {
+	if b.err != nil {
+		return b
+	}
+	if old := b.targetPeers.Get(p.GetStoreId()); old != nil {
+		b.err = errors.Errorf("cannot add peer %s: already have peer %s", p, old)
+	} else {
+		b.targetPeers.Set(p)
+	}
+	return b
+}
+
+// RemovePeer records a remove peer operation in Builder.
+func (b *Builder) RemovePeer(storeID uint64) *Builder {
+	if b.err != nil {
+		return b
+	}
+	if b.targetPeers.Get(storeID) == nil {
+		b.err = errors.Errorf("cannot remove peer from %d: not found", storeID)
+	} else if b.targetLeader == storeID {
+		b.err = errors.Errorf("cannot remove peer from %d: peer is target leader", storeID)
+	} else {
+		b.targetPeers.Delete(storeID)
+	}
+	return b
+}
+
+// PromoteLearner records a promote learner operation in Builder.
+func (b *Builder) PromoteLearner(storeID uint64) *Builder {
+	if b.err != nil {
+		return b
+	}
+	p := b.targetPeers.Get(storeID)
+	if p == nil {
+		b.err = errors.Errorf("cannot promote peer %d: not found", storeID)
+	} else if !p.GetIsLearner() {
+		b.err = errors.Errorf("cannot promote peer %d: not learner", storeID)
+	} else {
+		b.targetPeers.Set(&metapb.Peer{Id: p.GetId(), StoreId: p.GetStoreId()})
+	}
+	return b
+}
+
+// SetLeader records the target leader in Builder.
+func (b *Builder) SetLeader(storeID uint64) *Builder {
+	if b.err != nil {
+		return b
+	}
+	p := b.targetPeers.Get(storeID)
+	if p == nil {
+		b.err = errors.Errorf("cannot transfer leader to %d: not found", storeID)
+	} else if p.GetIsLearner() {
+		b.err = errors.Errorf("cannot transfer leader to %d: not voter", storeID)
+	} else {
+		b.targetLeader = storeID
+	}
+	return b
+}
+
+// SetPeers resets the target peer list.
+// If peer's ID is 0, the builder will allocate a new ID later.
+// If current target leader does not exist in peers, it will be reset.
+func (b *Builder) SetPeers(peers map[uint64]*metapb.Peer) *Builder {
+	if b.err != nil {
+		return b
+	}
+	b.targetPeers = peersMap{}
+	for _, p := range peers {
+		b.targetPeers.Set(p)
+	}
+	if _, ok := peers[b.targetLeader]; !ok {
+		b.targetLeader = 0
+	}
+	return b
+}
+
+// SetLightWeight marks the region as light weight. It is used for scatter regions.
+func (b *Builder) SetLightWeight() *Builder {
+	b.isLigthWeight = true
+	return b
+}
+
+// Build creates the Operator.
+func (b *Builder) Build(kind OpKind) (*Operator, error) {
+	if b.err != nil {
+		return nil, b.err
+	}
+
+	brief, err := b.prepareBuild()
+	if err != nil {
+		return nil, err
+	}
+
+	kind, err = b.buildSteps(kind)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewOperator(b.desc, brief, b.regionID, b.regionEpoch, kind, b.steps...), nil
+}
+
+// Initilize intermediate states.
+func (b *Builder) prepareBuild() (string, error) {
+	var voterCount int
+	for _, p := range b.targetPeers.m {
+		if !p.GetIsLearner() {
+			voterCount++
+		}
+	}
+	if voterCount == 0 {
```

Comment:
Cannot support to add a learner now? but we can use `{StoreID: 4}, {StoreID:5, IsLearner}` to pass the check.

Reply:
`targetPeers` is the region's peers after executing all commands, not the commands.

Reply:
Yes, I mean to `added a peer and a learner` after executing all commands.

Reply:
got.